src/Prefetch.cpp

```diff
@@ -0,0 +1,348 @@
+        scope.push(op->name, in);
+        debug(dbg_prefetch5) << "Let scope.push(" << op->name << ")\n";
+        IRMutator::visit(op);
+        debug(dbg_prefetch5) << "Let scope.pop(" << op->name << ")\n";
+        scope.pop(op->name);
+    }
+
+    void visit(const LetStmt *op) {
+        Interval in = bounds_of_expr_in_scope(op->value, scope);
+        scope.push(op->name, in);
+        debug(dbg_prefetch5) << "LetStmt scope.push(" << op->name << ")\n";
+        IRMutator::visit(op);
+        debug(dbg_prefetch5) << "LetStmt scope.pop(" << op->name << ")\n";
+        scope.pop(op->name);
+    }
+
+    void visit(const Realize *op) {
+        rscope.push(op->name, 1);
+        debug(dbg_prefetch5) << "Realize push(" << op->name << ")\n";
+        IRMutator::visit(op);
+    }
+
+    void visit(const For *op) {
+        Stmt body = op->body;
+
+        string func_name = tuple_func(op->name);
+        string ivar_name = tuple_var(op->name);
+        vector<Prefetch> &prefetches = get_func(func_name).schedule().prefetches();
+
+        // Add loop variable to interval scope for any inner loop prefetch
+        Expr var = Variable::make(Int(32), op->name);
+        Interval prein(var, var);
+        scope.push(op->name, prein);
+        debug(dbg_prefetch5) << "For scope.push(" << op->name << ", (" << var << "))\n";
+
+        debug(dbg_prefetch4) << "For: " << op->name << " " << func_name << " " << ivar_name << "\n";
+        if (prefetches.empty()) {
+            debug(dbg_prefetch4) << " No prefetch directives in schedule\n";
+        } else {
+            debug(dbg_prefetch4) << " Found prefetch directive(s) in schedule\n";
+        }
+
+        body = mutate(body);
+
+        for (const Prefetch &p : prefetches) {
+            debug(dbg_prefetch4) << " Check prefetch ivar: " << p.var << " == " << ivar_name << "\n";
+            if (p.var == ivar_name) {
+                debug(dbg_prefetch4) << " Found directive matching " << ivar_name << "\n";
+                debug(dbg_prefetch1) << " " << func_name
+                                     << " prefetch(" << ivar_name << ", " << p.offset << ")\n";
+
+                // Add loop variable + prefetch offset to interval scope for box computation
+                Expr var = Variable::make(Int(32), op->name);
+                Interval prein(var + p.offset, var + p.offset);
+                scope.push(op->name, prein);
+                debug(dbg_prefetch5) << " For scope.push(" << op->name << ", ("
+                                     << var << " + " << p.offset << ", "
+                                     << var << " + " << p.offset << "))\n";
+
+                map<string, Box> boxes;
+                boxes = boxes_required(body, scope);
+                // TODO: Opt: prefetch the difference from previous iteration
+                //            to the requested iteration (2 calls to boxes_required)
+
+                vector<pair<string, Expr>> plets;  // Prefetch let assignments
```

Comment:
I think it would be better to just have prefetch_box return a single Stmt that is a Block containing all the necessary LetStmts and calls.